auth/auth.go

```diff
@@ -781,3 +831,49 @@ func (auth *Authenticator) RegisterNewUser(username, email string) (User, error)
 
 	return user, err
 }
+
+// UpdateUserOIDCRolesChannels saves the user's current OIDC roles/channels, and resulting implicit roles/channels,
+// to the database.
+func (auth *Authenticator) UpdateUserOIDCRolesChannels(user_ User, roles base.Set, channels base.Set) error {
+	updateRolesChannelsCallback := func(currentPrincipal Principal) (updatedPrincipal Principal, err error) {
+		user, ok := currentPrincipal.(User)
+		if !ok {
+			return nil, base.ErrUpdateCancel
+		}
+
+		nextSeq := user.Sequence() + 1
+		user.SetSequence(nextSeq)
+
+		// TODO: these calls to rebuildRoles/rebuildChannels are somewhat expensive (require access/role-access queries)
+		// and we could hit this more than once if we're inside a CAS-loop.
+
+		newChans := user.OIDCChannels()
+		if newChans == nil {
+			newChans = ch.TimedSet{}
+		}
+		if !newChans.Equals(channels) {
+			newChans.UpdateAtSequence(channels, nextSeq)
+			user.SetOIDCChannels(newChans, nextSeq)
+			if err := auth.rebuildRoles(user); err != nil {
+				return nil, fmt.Errorf("failed to rebuild roles for %v: %w", base.UD(user.Name()).Redact(), err)
+			}
+		}
+
+		newRoles := user.OIDCRoles()
+		if newRoles == nil {
+			newRoles = ch.TimedSet{}
+		}
+		if !newRoles.Equals(roles) {
+			newRoles.UpdateAtSequence(roles, nextSeq)
+			user.SetOIDCRoles(newRoles, nextSeq)
+			if err := auth.rebuildChannels(user); err != nil {
+				return nil, fmt.Errorf("failed to rebuild channels for %v: %w", base.UD(user.Name()).Redact(), err)
+			}
+		}
+
+		base.DebugfCtx(auth.LogCtx, base.KeyAuth, "Updating user %s OIDC roles/channels to: %v, %v", base.UD(user.Name()), base.UD(user.OIDCRoles()), base.UD(user.OIDCChannels()).Redact())
```

Comment:
base.UD(user.OIDCChannels()).Redact() looks like a relatively expensive operation that's going to be performed here even if debug logging isn't enabled.  

Is there a reason to not just have base.UD(user.OIDCChannels()) here?